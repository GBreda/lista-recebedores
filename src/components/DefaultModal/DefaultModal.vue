<template>
  <div v-if="showModal" class="modal">
    <div class="modal--overlay">
      <div class="modal__container">
        <slot name="header">
          <div class="modal__container__header" data-test-id="header">
            <img
              src="@/assets/images/dark-close-icon.png"
              alt="Fechar modal"
              @click="closeModal"
              data-test-id="close-modal"
            />
          </div>
        </slot>
        <slot name="body" data-test-id="body" />

        <slot name="footer" data-test-id="footer">
          <div class="modal__container__footer">
            <button class="button-outline" @click="closeModal" data-test-id="back-button">
              Voltar
            </button>
            <div class="modal__container__footer__edit-actions">
              <button
                class="button-danger modal__container__footer__delete-button"
                @click="remove"
                data-test-id="remove-button"
              >
                <img src="@/assets/images/trash-icon.png" alt="Excluir recebedor" />
              </button>
              <button class="button-primary" @click="confirm" data-test-id="confirm-button">
                Salvar
              </button>
            </div>
          </div>
        </slot>
      </div>
    </div>
  </div>
</template>

<script setup>
defineProps({
  showModal: {
    type: Boolean,
    default: false
  }
})

const emit = defineEmits(['close:modal', 'update:confirm', 'update:remove'])

const closeModal = () => {
  emit('close:modal')
}

const confirm = () => {
  emit('update:confirm')
}

const remove = () => {
  emit('update:remove')
}
</script>

<style lang="scss" scoped>
.modal {
  &--overlay {
    position: fixed;
    z-index: 9998;
    top: 0;
    left: 0;
    width: 100%;
    height: 100%;
    background: rgba($black, 0.2);
    backdrop-filter: blur(0.4rem);
  }

  &__container {
    width: 73rem;
    margin: 4rem auto;
    padding: 2.5rem 2.5rem 4rem 2.5rem;
    background-color: $white;
    border-radius: 0.2rem;
    box-shadow: 0 0.1rem 0.5rem 0 rgba($black, 0.1);

    &__header {
      display: flex;
      justify-content: flex-end;

      img {
        cursor: pointer;
      }
    }

    &__footer {
      margin-top: 3.5rem;
      display: flex;
      justify-content: space-between;

      &__edit-actions {
        display: flex;
        align-items: center;
        gap: 2rem;
      }

      &__delete-button {
        width: 6rem;
      }
    }
  }
}
</style>
