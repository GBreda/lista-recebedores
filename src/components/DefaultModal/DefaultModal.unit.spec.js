import { describe, it, expect } from 'vitest'
import { mount } from '@vue/test-utils'

import DefaultModal from './DefaultModal.vue'

const mountDefaultModal = ({ slots } = {}) => {
  return mount(DefaultModal, {
    props: {
      showModal: true
    },
    slots: {
      ...slots
    }
  })
}

describe('DefaultModal', () => {
  let wrapper

  it('should mount the component', () => {
    wrapper = mountDefaultModal()

    expect(wrapper.vm).toBeDefined()
  })

  it('should match the snapshot', () => {
    wrapper = mountDefaultModal()

    expect(wrapper.element).toMatchSnapshot()
  })

  it('should emit close event on close button click', async () => {
    wrapper = mountDefaultModal()

    await wrapper.find("[data-test-id='close-modal']").trigger('click')

    expect(wrapper.emitted()['close:modal']).toBeTruthy()
  })

  it('should emit close event on back button click', async () => {
    wrapper = mountDefaultModal()

    await wrapper.find("[data-test-id='back-button']").trigger('click')

    expect(wrapper.emitted()['close:modal']).toBeTruthy()
  })

  it('should emit remove event on remove button click', async () => {
    wrapper = mountDefaultModal()

    await wrapper.find("[data-test-id='remove-button']").trigger('click')

    expect(wrapper.emitted()['update:remove']).toBeTruthy()
  })

  it('should emit confirm event on confirm button click', async () => {
    wrapper = mountDefaultModal()

    await wrapper.find("[data-test-id='confirm-button']").trigger('click')

    expect(wrapper.emitted()['update:confirm']).toBeTruthy()
  })

  it('should render header slot content', () => {
    const wrapper = mountDefaultModal({
      slots: {
        header: '<div data-test-id="header">Custom Header Content</div>'
      }
    })

    expect(wrapper.find('[data-test-id="header"]').text()).toBe('Custom Header Content')
  })

  it('should render body slot content', () => {
    const wrapper = mountDefaultModal({
      slots: {
        body: '<div data-test-id="body">Custom Body Content</div>'
      }
    })

    expect(wrapper.find('[data-test-id="body"]').text()).toBe('Custom Body Content')
  })

  it('should render footer slot content', () => {
    const wrapper = mountDefaultModal({
      slots: {
        footer: '<div data-test-id="footer">Custom Footer Content</div>'
      }
    })

    expect(wrapper.find('[data-test-id="footer"]').text()).toBe('Custom Footer Content')
  })
})
