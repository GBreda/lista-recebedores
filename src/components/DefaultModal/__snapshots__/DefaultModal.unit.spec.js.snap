// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`DefaultModal > should match the snapshot 1`] = `
<div
  class="modal"
  data-v-5db2f97c=""
>
  <div
    class="modal--overlay"
    data-v-5db2f97c=""
  >
    <div
      class="modal__container"
      data-v-5db2f97c=""
    >
      
      <div
        class="modal__container__header"
        data-test-id="header"
        data-v-5db2f97c=""
      >
        <img
          alt="Fechar modal"
          data-test-id="close-modal"
          data-v-5db2f97c=""
          src="/src/assets/images/dark-close-icon.png"
        />
      </div>
      
      
      
      
      <div
        class="modal__container__footer"
        data-v-5db2f97c=""
      >
        <button
          class="button-outline"
          data-test-id="back-button"
          data-v-5db2f97c=""
        >
           Voltar 
        </button>
        <div
          class="modal__container__footer__edit-actions"
          data-v-5db2f97c=""
        >
          <button
            class="button-danger modal__container__footer__delete-button"
            data-test-id="remove-button"
            data-v-5db2f97c=""
          >
            <img
              alt="Excluir recebedor"
              data-v-5db2f97c=""
              src="/src/assets/images/trash-icon.png"
            />
          </button>
          <button
            class="button-primary"
            data-test-id="confirm-button"
            data-v-5db2f97c=""
          >
             Salvar 
          </button>
        </div>
      </div>
      
    </div>
  </div>
</div>
`;
